packages/runtime/types/locale.ts

```diff
@@ -0,0 +1,471 @@
+import { z } from 'zod'
+
+export const localeSchema = z.string().transform((value, ctx) => {
+  const { language, region } = new Intl.Locale(value)
+
+  if (LANGUAGE_CODE_SET.has(language as LanguageCode) && region == null && value === language) {
+    return value as Locale
+  }
+
+  if (
+    LANGUAGE_CODE_SET.has(language as LanguageCode) &&
+    COUNTRY_CODE_SET.has(region as CountryCode) &&
+    value === `${language}-${region}`
+  ) {
+    return value as Locale
+  }
+
+  ctx.addIssue({
+    code: z.ZodIssueCode.custom,
+    message: `Locale is not supported: '${value}'. Make sure the locale is in the format of 'language', e.g. 'en', or 'language-country', e.g. 'en-US'.`,
+  })
+
+  return z.NEVER
+})
```

Comment:
Why not have the transform return an `Intl.Locale`? It seems weird to cast to a special `Locale` type when we can use the native `Intl.Locale`

Reply:
I was thinking about that too. I remember that we could just use `Intl.Locale` everywhere. I decided to just use string because that's what we use in the `State.Locale`. Do you also want to store `Intl.Locale` in the state?

Reply:
We need to keep our domains separate. Whether we use a `string` or whatever else in the state shouldn't necessarily drive what's in the public API. But yeah, I don't see why we can't just keep an `Intl.Locale` in state unless we rely on it being serializable. Regardless, it would be very easy to keep the serialized string representation in state and then have a selector that transforms it into a `Intl.Locale` whenever read from state

Reply:
I can see that we can use `Intl.Locale` internally when reading from the state.

The custom `Locale` type here is used mainly to type the input in the `ReactRuntime`:
```
export const runtime = new ReactRuntime({
  unstable_i18n: {
    locales: ["es", "en-US", "es-SE", "id-ID"],
    defaultLocale: "en-US",
  },
});
```

Reply:
That type has nothing to do with `localeSchema`, though. It's just defined as a combination of the language code and country code enums, which can be used at the input of the API and should still be independent of what's stored internally.

The values allowed as inputs can follow the locale string type while internally we can always use `Intl.Locale`

Reply:
I updated the custom locale string to be just used when parsing the input from the user, and I made it so that we use `Intl.Locale` on the selector.

We still need to use `string` on the actions, because we're going to send it through the `postMessage`.

Reply:
Yes, yes. I called it `LocaleInput` though. I can change it if needed.

Reply:
Let's change it to `LocaleString` since it's not necessarily just an input (e.g., it can be used as an output in certain cases)

Reply:
Updated